### PR TITLE
Ensure Proxy of non-constructible object is not constructible fixes #4724

### LINKED ISSUE #4724: Proxy Object can be constructible when its target was not
Per spec if the target of a Proxy object is not constructible then the Proxy object should not have a construct method i.e. should not be constructible even if supplied with a construct method.

See Note 1 here: [#sec-proxy-object-internal-methods-and-internal-slots-construct-argumentslist-newtarget](https://tc39.github.io/ecma262/#sec-proxy-object-internal-methods-and-internal-slots-construct-argumentslist-newtarget)
And 7. b) here: [#sec-proxycreate](https://tc39.github.io/ecma262/#sec-proxycreate)

This doesn't work in CC - any Proxy created in CC if supplied with a construct method will be callable with new:

```js
let obj = new Proxy(()=>{}, {construct: ()=>({})});
new obj();//expected error -> actual no error
```

```js
let obj = new Proxy(async function foo(){}, {construct: ()=>({})});
new obj();//expected error -> actual no error
```

Both of these cases error as I'd expect in v8.

EDIT: I've made a PR to fix this.

### PR DESCRIPTION
**Pull request by rhuanjl**
A rather rough/quick fix for issue #4724 which I just raised.

Not sure if this is best fix but it seems to work

### CODE REVIEW

File: lib/Runtime/Library/JavascriptProxy.cpp [hunk 1]
```diff
@@ -2191,6 +2191,11 @@ namespace Js
         varArgs[1] = targetObj;
         if (args.Info.Flags & CallFlags_New)
         {
+            if(!JavascriptOperators::IsConstructor(targetObj))
```
Comment (boingoing) [inline lib/Runtime/Library/JavascriptProxy.cpp]: Nitpick: space after if

File: lib/Runtime/Library/JavascriptProxy.cpp [hunk 2]
```diff
@@ -2039,7 +2039,7 @@ namespace Js
     Var JavascriptProxy::ConstructorTrap(Arguments args, ScriptContext* scriptContext, const Js::AuxArray<uint32> *spreadIndices)
     {
         PROBE_STACK(GetScriptContext(), Js::Constants::MinStackDefault);
-
+
```
Comment (boingoing) [inline lib/Runtime/Library/JavascriptProxy.cpp]: Nitpick: unnecessary whitespace

File: test/es6/proxybugs.js [hunk 3]
```diff
@@ -378,11 +378,25 @@ var tests = [
...
-            var obj = Proxy.revocable(() => {}, handler);
+            var obj = Proxy.revocable(function() {}, handler);
             new obj.proxy();
             assert.isTrue(trapCalled);
         }
     },
+    {
+        name: "Assertion validation : proxy of arrow function should not be constructable",
+        body() {
+            let obj = new Proxy(()=>{}, {construct: ()=>({})});
```
Comment (boingoing) [inline test/es6/proxybugs.js]: We should test a few more cases like non-constructible builtins (Math.abs, eval, parseInt, etc), 

class members:
```js
class a { 
  static foo() { } 
  bar() { } 
};
new a.foo();
new (new a()).bar();
```
object literal methods:
```js
var obj = { foo() { } };
new obj.foo();
```

File: lib/Runtime/Library/JavascriptProxy.cpp (hunk 2 above)
Comment (jackhorton) [inline lib/Runtime/Library/JavascriptProxy.cpp]: Turbo-nit: added spaces.

**Comment by jackhorton** [comment]
@Cellule is this OSX WASM issue known?

**Comment by rhuanjl** [comment]
@boingoing points addressed and updated - note I added a lot of test cases so I made a new file for them.

Also the wasm OS X test hasn’t failed after I re-pushed considering that this change is nothing to do with wasm assume it’s just a flaky/slow test.

File: test/es6/proxyconstruction.js [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        }
+    },
+    {
+        name: "Proxy of normal function should be constructable",
+        body() {
+            testConstructProxy(function test6 () {}, false);
+        }
+    },
+    {
+        name: "Proxy of static class member function should be constructable",
```
Comment (boingoing) [inline test/es6/proxyconstruction.js]: nitpick: title says this should be constructable but we mean it should "not be" constructable, right?

File: test/es6/proxyconstruction.js [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+        body() {
+            class testing1
+            {
+                static test () {}
+            }
+            testConstructProxy(testing1.test, true);
+        }
+    },
+    {
+        name: "Proxy of class member function should be constructable",
```
Comment (boingoing) [inline test/es6/proxyconstruction.js]: nitpick: as above

**Comment by boingoing** [comment]
@rhuanjl Thanks for expanding on the test cases, these are great. Please rebase on top of master and I can help land this one.

**Comment by rhuanjl** [comment]
@boingoing it got out of date after I rebased so I rebased again, please let me know if you need me to do anything else.

**Comment by boingoing** [comment]
@rhuanjl This looks good to me, I have it ready to go from our internal-side as well. Just fighting with a piece of internal infrastructure and I'll pull the change.

**Comment by boingoing** [comment]
@rhuanjl Got it. Thanks very much for your contribution!

**Pull request merged**